PowerShell Loops and Iteration Module - closes #173 (#180)
bob-crutchley/courseware: topics/powershell/modules/loops-and-iteration/README.md

# Loops and Iteration
<!--TOC_START-->
### Contents
- [Overview](#overview)
- [For Loop](#for-loop)
	- [Iterating an Array](#iterating-an-array)
- [While Loop](#while-loop)
	- [Iterating an Array](#iterating-an-array-1)
	- [Waiting for a File to be Deleted](#waiting-for-a-file-to-be-deleted)
	- [Infinite Loops](#infinite-loops)
- [Do While Loops](#do-while-loops)
	- [Do While Loops for User Input](#do-while-loops-for-user-input)
- [Do Until Loops](#do-until-loops)
	- [Waiting for a File to Exist](#waiting-for-a-file-to-exist)
- [Tasks](#tasks)
	- [File Contents](#file-contents)
	- [For Loop](#for-loop-1)
	- [While Loop](#while-loop-1)
	- [Do While Loop](#do-while-loop)
	- [Do Until Loop](#do-until-loop)

<!--TOC_END-->
## Overview

> Arrays and iteration can be used in PowerShell for solving extensive, repetetive, procedural tasks.

Reviewer: ```suggestion
Arrays and iteration can be used in PowerShell for solving extensive, repetitive, procedural tasks.
```

One of PowerShells main purpose is to automate tasks on Windows, often for multiple machines.

The way that arrays can be iterated is by using loops.

> There are a few different types of loops which can utilise depending on the situation.

Reviewer: ```suggestion
There are a few different types of loops which can be utilised depending on the situation.
```

## For Loop

> The For loop is a very bsaic type of loop which can be used in most situations.

Reviewer: ```suggestion
The For loop is a very basic type of loop which can be used in most situations.
```

}
```
### Iterating an Array
It would be slightly better to use a For loop to iterate arrays but it can still be done with a While loop:
```powershell
$colours = @("Red", "Blue", "Green", "Orange", "Purple")
While ($i -ne $colours.length) {
    $colours[$i]
}
```
### Waiting for a File to be Deleted
A more appropriate implementation of a while loop would be to use it on something that needs to wait until a condition is met, such as a file being deleted.

In the example shown below the following is evaluated:
- Set a `filePath` variable for a file called `file.txt` in the user's home directory
- For the loop condition check the file exists with `Test-Path`, this will return as **True** when the file exists
```powershell
$filePath = "$HOME\file.txt"
While (Test-Path "$filePath") {
    "Waiting for $filePath to be deleted exist..."
    Start-Sleep 2
}
```
### Infinite Loops
Something to be very cautious of when using While loops (and any type of loop for that matter) is to not get into an infinite loop.
If the condition for the loop is always met, then it will never stop running.

Taking the waiting for a file example, we can add a counter and a second condition to the loop.

> For example here, if the loop has executed 10 times, then it will exit because both conditions inside the loop condition need to be met:

Reviewer: Shouldn't this read something more along the lines of: "this loop will only exit when both conditions have been met, i.e once the file has been deleted and the counter reaches 10" - its a little confusing I think.

Also I dont see what this has to do with Infinite loops? have they discussed -and already? is there a portion on logical operators?

Author: As a while loop it will continue running condition is not met, so while:
- The file exists
- The counter is less than 10
The loop will continue to run, I can update this part,  clarifying the explanation better like this I think.

I've opened an issue for a module to cover logical and relational operators: #284

```powershell
$counter = 0
$filePath = "$HOME\file.txt"
While (Test-Path "$filePath" -and ($counter -lt 10)) {
    "Waiting for $filePath to be deleted..."
    Start-Sleep 1
    $counter++
}
```
## Do While Loops

> These are very similar to the while loop with once exception, the loops command block is executed before the condition is evaluted.

Reviewer: ```suggestion
These are very similar to the while loop with one exception, the loops command block is executed before the condition is evaluated.
```

The syntax of a Do Until loop is strucured the same as a Do While loop:
```powershell
Do {
    # commands
} Until([CONDITION])
```
### Waiting for a File to Exist
Because the condition for this type of loop is flipped, we can take the example from before to wait for file to be deleted and use this to wait for a file to be created:
```powershell
Do {
    "Waiting for $filePath to be deleted..."
    Start-Sleep 1
} Until (Test-Path "$filePath") 
```
## Tasks
We are going to use the different types of loops explained above to iterate over a log file (`log.txt`) and extract the errors, which can be found in this folder.
While working through this set of tasks make sure that you are creating and running the scripts in this directory.
### File Contents
Before we can try this out with any type of loop we will need to get the file content to iterate through.
So for each of the following steps you will need this at the top of your script:
```powershell
$content = Get-Content "log.txt"
```
### For Loop
With the for loop we can iterate over the file content just like an array, checking if the element contains an error message, printing it out to the console if it does. 

> Create a file in this folder called `error-check-for-loop.ps1` in this directory and enter the following:

Reviewer: ```suggestion
Create a file in this folder called `error-check-for-loop.ps1` and enter the following:
```

```powershell
$content = Get-Content "log.txt"
For ($i=0; $i -lt $content.Length; $i++) {
    if ($content[$i].Contains("[error]")) {
        "Line: $($i+1): $($content[$i])"
    }
}
```
Now try executing the script:
```powershell

> ./error-check-for-loop.ps1

Reviewer: Don't laugh! However I have a authorization error running the script, hopefully this is something that is resolved in a previous session, someone like me jumping in at this point could get the same problem (yes I ran as admin - apparently scripts are disabled on my system)

Author: This is to do with execution policies on Windows for PowerShell, looks like there are blocked in some cases.

Scripts can be allowed by running this in a PowerShell session run as an Admin:
```powershell
Set-ExecutionPolicy -ExecutionPolicy Unrestricted
```
I'd say we should have a whole module on this because this issue would come up a lot on site I can see, for now though it might be worth putting that one-liner in the introduction to PowerShell or something.
